Mathlib/Algebra/Order/Module/Archimedean.lean

```diff
@@ -0,0 +1,149 @@
+/-
+Copyright (c) 2025 Weiyi Wang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Weiyi Wang
+-/
+
+import Mathlib.Algebra.Order.Archimedean.Class
+import Mathlib.Algebra.Order.Module.Basic
+import Mathlib.LinearAlgebra.DFinsupp
+
+/-!
+# Further lemmas related to Archimedean classes for ordered module.
+
+## Main definitions
+* `ArchimedeanClass.ball` and `ArchimedeanClass.closedBall`:
+  `ArchimedeanClass.ballAddSubgroup` and `ArchimedeanClass.closedBallAddSubgroup` as submodules.
+* `ArchimedeanClass.IsGrade`: The module version of `ArchimedeanClass.IsGradeAddSubgroup`.
+-/
+
+namespace ArchimedeanClass
+
+variable {M : Type*}
+variable [AddCommGroup M] [LinearOrder M] [IsOrderedAddMonoid M]
+variable {K : Type*} [Ring K] [LinearOrder K] [IsOrderedRing K] [Archimedean K]
+variable [Module K M] [PosSMulMono K M]
+
+variable (K) in
+/-- For a ordered module over an `Archimedean` ring,
+`ArchimedeanClass.addSubgroup` is a submodule. -/
+noncomputable
+def submodule (s : UpperSet (ArchimedeanClass M)) : Submodule K M where
+  __ := addSubgroup s
+  smul_mem' k {a} := by
+    obtain rfl | hs := eq_or_ne s ⊤
+    · aesop
+    · change a ∈ addSubgroup s → k • a ∈ addSubgroup s
+      simp_rw [mem_addSubgroup_iff hs]
+      intro ha
+      obtain _ | _ := subsingleton_or_nontrivial K
+      · rw [Subsingleton.eq_zero k]
+        simpa using (IsUpperSet.top_mem s.upper).mpr <| UpperSet.coe_nonempty.mpr hs
+      · obtain ⟨n, hn⟩ := Archimedean.arch |k| (show 0 < 1 by simp)
+        refine s.upper (mk_le_mk.mpr ⟨n, ?_⟩) ha
+        have : n • |a| = (n • (1 : K)) • |a| := by rw [smul_assoc, one_smul]
+        rw [this, abs_smul]
+        exact smul_le_smul_of_nonneg_right hn (by simp)
```

Comment:
The line of sight rule in usual programming says that the main branch of a function should remain unindented. I find this very useful for proofs also:
```suggestion
    obtain rfl | hs := eq_or_ne s ⊤
    · aesop
    change a ∈ addSubgroup s → k • a ∈ addSubgroup s
    simp_rw [mem_addSubgroup_iff hs]
    intro ha
    obtain _ | _ := subsingleton_or_nontrivial K
    · rw [Subsingleton.eq_zero k]
      simpa using (IsUpperSet.top_mem s.upper).mpr <| UpperSet.coe_nonempty.mpr hs
    obtain ⟨n, hn⟩ := Archimedean.arch |k| (show 0 < 1 by simp)
    refine s.upper (mk_le_mk.mpr ⟨n, ?_⟩) ha
    have : n • |a| = (n • (1 : K)) • |a| := by rw [smul_assoc, one_smul]
    rw [this, abs_smul]
    exact smul_le_smul_of_nonneg_right hn (by simp)
```